[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Bingjing-Dong/CSC448-AI/blob/main/Assignment%203.ipynb)

In [1]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
dataset = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
dataset.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset[['Temperature °C','Mols KCL']]
y = dataset['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

## Part 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = np.array([[294, 776]])
print("Prediction: " + str(reg.predict(sample)))
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("Score: " + str(reg.score(X_test,y_test)))
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficent: " + str(reg.coef_))
print("Intercept: " + str(reg.intercept_))

Prediction: [639732.83190002]
Score: 0.858202107053349
Coefficent: [ 877.26717297 1026.45679272]
Intercept: -414714.18810411997


The score function return the coefficient of determination of the prediction. Our score is 0.8582 here, which means the model has 86.161% accuracy of predicting the slime size given the temperature and mols.

The equation of $h(x) = 877.26717x_1 + 1026.45679x_2 - 414714.1881$

## Part 4. Use Cross Validation

In [6]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(reg, X, y, cv=5)
# Report on their finding and their significance
print("Cross-validation Score: " + str(scores))
print("%0.5f accuracy with a standard deviation of %0.5f" % (scores.mean(), scores.std()))

Cross-validation Score: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
0.85682 accuracy with a standard deviation of 0.01347


The simplest way to use cross-validation is to call the cross_val_score helper function on the estimator and the dataset.

The accuracy of fitting the model and computing the score 5 consecutive times (with different splits each time) is between 0.83912 and 0.87203. The mean accuracy is 0.85682 which is closed to the test data score above.

## Part 5. Using Polynomial Regression

In [7]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
# Report on the metrics and output the resultant equation as you did in Part 3.
X_train, X_test, y_train, y_test = train_test_split(poly.fit_transform(X), y, test_size=0.1)
poly_reg = LinearRegression().fit(X_train, y_train)
print("Score: " + str(poly_reg.score(X_test,y_test)))
print("coefficent: " + str(poly_reg.coef_))
print("intercept: " + str(poly_reg.intercept_))

Score: 1.0
coefficent: [ 0.00000000e+00  1.20000000e+01 -1.39313366e-07 -3.57011649e-11
  2.00000000e+00  2.85714287e-02]
intercept: 1.6176840290427208e-05


After we use PolynomialFeatures library with data, the score became 1, which means the train model using polynomial regression has 100% accurate.

The equation of $h(x) = 0.00002 + 0.12x_1 + 2x_1x_2 + 0.02857x_2^2$